chain/xrp/client/client.go

```diff
@@ -0,0 +1,816 @@
+	OwnerCount        int    `json:"OwnerCount"`
+	PreviousTxnID     string `json:"PreviousTxnID"`
+	PreviousTxnLgrSeq int64  `json:"PreviousTxnLgrSeq"`
+	Sequence          int64  `json:"Sequence"`
+	Index             string `json:"Index"`
+}
+
+func (client *Client) FetchBaseInput(ctx context.Context, args xcbuilder.TransferArgs) (xrptxinput.TxInput, error) {
+	txInput := xrptxinput.NewTxInput()
+
+	account := args.GetFrom()
+
+	currentSequence, err := client.getNextValidSeqNumber(account)
+	if err != nil {
+		return xrptxinput.TxInput{}, err
+	}
+	currentSequencePtr := *currentSequence
+	txInput.Sequence = currentSequencePtr
+
+	ledgerSequence, err := client.getLatestValidatedLedgerSequence()
+	if err != nil {
+		return xrptxinput.TxInput{}, err
+	}
+	ledgerSequencePtr := *ledgerSequence
+	ledgerOffset := int64(20) // Ledger offset
+	lastLedgerSequence := ledgerSequencePtr + ledgerOffset
+	txInput.LastLedgerSequence = lastLedgerSequence
+
+	return *txInput, nil
+}
+
+// FetchTransferInput returns tx input for a Template tx
+func (client *Client) FetchTransferInput(ctx context.Context, args xcbuilder.TransferArgs) (xc.TxInput, error) {
+	txInput, err := client.FetchBaseInput(ctx, args)
+	if err != nil {
+		return nil, err
+	}
+
+	return &txInput, nil
+}
+
+// Deprecated method - use FetchTransferInput
+func (client *Client) FetchLegacyTxInput(ctx context.Context, from xc.Address, to xc.Address) (xc.TxInput, error) {
+	// No way to pass the amount in the input using legacy interface, so we estimate using min amount.
+	args, _ := xcbuilder.NewTransferArgs(from, to, xc.NewAmountBlockchainFromUint64(1))
+	txInput, err := client.FetchTransferInput(ctx, args)
+	if err != nil {
+		return nil, err
+	}
+
+	return txInput, nil
+}
+
+// SubmitTx submits a Template tx
+func (client *Client) SubmitTx(ctx context.Context, txInput xc.Tx) error {
+	serializedTxInputBytes, err := txInput.Serialize()
+	if err != nil {
+		return err
+	}
+
+	serializedTxInputHex := hex.EncodeToString(serializedTxInputBytes)
+	serializedTxInputHexBytes := []byte(serializedTxInputHex)
+
+	submitRequest := &SubmitRequest{
+		Method: "submit",
+		Params: []SubmitParamEntry{
+			{
+				TxBlob: string(serializedTxInputHexBytes),
+			},
+		},
+	}
+
+	var submitResponse SubmitResponse
+	err = client.Send(MethodPost, submitRequest, &submitResponse)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// FetchTxInfo Returns transaction info - new endpoint
+func (client *Client) FetchTxInfo(ctx context.Context, txHash xc.TxHash) (xclient.TxInfo, error) {
+	legacyTxInfo, err := client.FetchLegacyTxInfo(ctx, txHash)
+	if err != nil {
+		return xclient.TxInfo{}, err
+	}
+
+	// Remap to new tx
+	return xclient.TxInfoFromLegacy(client.Asset.GetChain().Chain, legacyTxInfo, xclient.Account), nil
+}
+
+// FetchLegacyTxInfo Returns transaction info - legacy/old endpoint
+func (client *Client) FetchLegacyTxInfo(ctx context.Context, txHash xc.TxHash) (xc.LegacyTxInfo, error) {
```

Comment:
The logic in here regarding parsing the metadata to figure out the transfer flows isn't super clear to me and gets a bit nested.  I'd appreciate some helper functions to simplify (I added an example), and also some comments that explain how it works, at least at a high level.